### PR TITLE
3D Generation: InstantMesh Training, Inference, and Eval Scripts

### PR DESCRIPTION
**Pull request by HaFred**
This PR implements [InstantMesh](https://github.com/TencentARC/InstantMesh) for 3D meshing using multiview images. The torch-cuda version with textures is shown below.

images

Input is `nx3xhxw` multiview video and the output is a 3D mesh `mesh.obj` file with vertices, faces.
HW environment specs are as follows.
```
Mindspore Version:               2.3.1 release
CANN Version:                    7.3
Ascend Driver:                   23.0.rc3.6
```

## 3D Mesh Demos
Inference demos with the pretrained checkpoint can be found in the readme, with any HTML renderer such as VSCode it can be viewed and interacted with smoothly as shown below.
image

These links can be found here for easy access.
* [akun](https://sketchfab.com/models/c8b5b475529d48589b85746aab638d2b/embed)
* [anya](https://sketchfab.com/models/180fd247ba2f4437ac665114a4cd4dca/embed)

Input multiview images are illustrated [here](https://github.com/HaFred/mindone/blob/itmh_jul26/examples/sv3d/README.md), respectively. Kindly notice that the input multiview images can be either retrieved viav the [SV3D pipeline](https://github.com/mindspore-lab/mindone/pull/574) or Zero123++ pipeline as in the original implementation, the paper's core contribution is the process of 3D meshing out of multiview images.

### 1. Cache Miss in `mint.unique()` and other ACLNN Operators
As introduced in the [Readme](https://github.com/HaFred/mindone/blob/itmh_jul26/examples/sv3d/models/instantmesh/README.md), InstantMesh extracts isosurface for meshing using [FlexiCubes](https://github.com/TencentARC/InstantMesh/blob/main/src/models/geometry/rep_3d/flexicubes.py#L303), which essentially requires a _unique_  operation to determine the surfaces of an object. Unfortunately, the operator `mint.unique()` remains not supported by ACLNN but only AICPU, and hence leads to the program being stuck by cache misses, as told by the MindSpore framework colleagues. Once the `mint.unique()` is fully supported by A+M, we will implement the 3D meshing with FlexiCubes for higher resolution.

(**Update on August 1st**: Turns out the CANN operator `aclnnUniqueDim` takes too long. Interacting with the framework and CANN team for solutions.)

### Workaround: 3D Meshing with the Raw Triplane Features using Marching Cubes
For the reasons above, we have to find a workaround to extract 3D meshes from multiview images (from SV3D in our case). Here in this PR, since we already have a rough SDF extracted from the SDF MLP heads with the triplane features input, a straightforward way is to take the rough SDF and feed it to the classic isosurface extraction method, such as the Marching Cubes. As the optimization in the Marching Cubes lacks degrees of freedom to represent high-quality meshes, it tends to use more vertices (naturally more faces, triangles) to fit an irregular 3D shape, especially when the 3D shape cannot be approximated to a surface. Details can be found in [the FlexiCubes' paper](https://arxiv.org/abs/2308.05371).

| FlexiCubes | Marching Cubes|
| --- | --- |
| image    | image|

### 2. CUDA Extension for Rasterization
InstantMesh uses [nvdiffrast](https://github.com/NVlabs/nvdiffrast) for uv map rasterization and 2D rendering for the FlexiCube 3D volume. Bypass this for now.

### CODE REVIEW

File: examples/instantmesh/README.md [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+| <div class="sketchfab-embed-wrapper"><iframe title="akun_ms" frameborder="0" allowfullscreen mozallowfullscreen="true" webkitallowfullscreen="true" allow="autoplay; fullscreen; xr-spatial-tracking" xr-spatial-tracking execution-while-out-of-viewport execution-while-not-rendered web-share src="https://sketchfab.com/models/c8b5b475529d48589b85746aab638d2b/embed"></iframe></div> | <div class="sketchfab-embed-wrapper"><iframe title="anya_ms" frameborder="0" allowfullscreen mozallowfullscreen="true" webkitallowfullscreen="true" allow="autoplay; fullscreen; xr-spatial-tracking" xr-spatial-tracking execution-while-out-of-viewport execution-while-not-rendered web-share src="https://sketchfab.com/models/180fd247ba2f4437ac665114a4cd4dca/embed"></iframe></div> |
+
+The illustrations here are better viewed in viewers than with HTML support (e.g., the vscode built-in viewer).
+
+## Environments
```
Comment (SamitHuang) [inline examples/instantmesh/README.md:71]: Pls refactor this section referring to 
https://github.com/mindspore-lab/mindone/wiki/%E6%A8%A1%E5%9E%8B%E7%89%88%E6%9C%AC%E4%B9%A6%E5%86%99%E6%A0%B7%E4%BE%8B

File: examples/instantmesh/train.py [hunk 2]
```diff
@@ -0,0 +1,460 @@
+""" InstantMesh Stage-1 Training Script """
```
Comment (CaitinZhao) [inline examples/instantmesh/train.py:1]: rename to train_stage1.py?

File: mindone/trainers/callback.py [hunk 3]
```diff
@@ -267,7 +267,9 @@ def on_train_epoch_end(self, run_context):
                     self.ema.swap_before_eval()
 
                 # save history checkpoints
-                self.ckpt_manager.save(self.net_to_save, None, ckpt_name=ckpt_name, append_dict=append_dict)
+                self.ckpt_manager.save(
+                    self.net_to_save, perf=cb_params["net_outputs"], ckpt_name=ckpt_name, append_dict=append_dict
```
Comment (SamitHuang) [inline mindone/trainers/callback.py:271]: why add perf=cb_params["net_outputs"]? does it fit for other models?

File: examples/instantmesh/train.py (hunk 2 above)
Comment (HaFred) [reply examples/instantmesh/train.py]: The `train.py` is for both stage 1 and stage 2 training. The difference in stage 2 is in the renderer side, where a flexicube trainable isosurface extraction is included to incorporate additional depth, normal, and regularization loss. Due to the incapability of operators like `mint.unique()` and cuda extensions, stage 2 is not included yet.

Comment (HaFred) [reply examples/instantmesh/train.py]: This confusing comment snippet is removed. Thanks.

File: examples/instantmesh/README.md [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+2. Inference is tested on the machine with the following specs using 1x NPU:
+
+| mindspore |	ascend driver | firmware	| cann toolkit/kernel |
+| :--- | :--- | :--- | :--- |
+| 2.3.1	    | 24.1.RC2 |7.3.0.1.231	| 8.0.RC2.beta1 |
+
+## Pretrained Models
+### ViT Pretrained Checkpoint
+To better accommodate the mindone transformer codebase, we provide an out-of-the-box [checkpoints conversion script](./tools/convert_dinovit_bin2st.py) that works seamlessly with the mindspore version of transformers.
```
[the remark this answers is not in the document]
Comment (HaFred) [reply examples/instantmesh/README.md:84]: Done

File: examples/instantmesh/README.md [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+## Training
+```shell
+python train.py --base configs/YOUR_CFG
+```
+One needs to patch `mindcv.models.vgg` in L62 to enable conv kernel bias to align with the torchmetric implementation of lpips loss.
+```diff
+- conv2d = nn.Conv2d(in_channels, v, kernel_size=3, pad_mode="pad", padding=1)
++ conv2d = nn.Conv2d(in_channels, v, kernel_size=3, pad_mode="pad", padding=1, has_bias=True)
+```
+
```
[the remark this answers is not in the document]
Comment (HaFred) [reply examples/instantmesh/README.md:113]: Done

File: mindone/trainers/callback.py (hunk 3 above)
Comment (HaFred) [reply mindone/trainers/callback.py:281]: For all current examples, eval callback is setup as `latest_k`. However, I found that this model's latest trained ckpt is not necessarily good. Thus, I passed in `top_k` to the callback, which requires loss to be recorded under this line. This modification does not affect `latest_k` setup in other examples.

File: examples/instantmesh/README.md [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+## Training
+```shell
+python train.py --base configs/YOUR_CFG
+```
+One needs to patch `mindcv.models.vgg` in L62 to enable conv kernel bias to align with the torchmetric implementation of lpips loss.
+```diff
+- conv2d = nn.Conv2d(in_channels, v, kernel_size=3, pad_mode="pad", padding=1)
++ conv2d = nn.Conv2d(in_channels, v, kernel_size=3, pad_mode="pad", padding=1, has_bias=True)
+```
```
[the remark this answers is not in the document]
Comment (HaFred) [reply examples/instantmesh/README.md:112]: done

File: examples/instantmesh/README.md [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+| <div class="sketchfab-embed-wrapper"><iframe title="akun_ms" frameborder="0" allowfullscreen mozallowfullscreen="true" webkitallowfullscreen="true" allow="autoplay; fullscreen; xr-spatial-tracking" xr-spatial-tracking execution-while-out-of-viewport execution-while-not-rendered web-share src="https://sketchfab.com/models/c8b5b475529d48589b85746aab638d2b/embed"></iframe></div> | <div class="sketchfab-embed-wrapper"><iframe title="anya_ms" frameborder="0" allowfullscreen mozallowfullscreen="true" webkitallowfullscreen="true" allow="autoplay; fullscreen; xr-spatial-tracking" xr-spatial-tracking execution-while-out-of-viewport execution-while-not-rendered web-share src="https://sketchfab.com/models/180fd247ba2f4437ac665114a4cd4dca/embed"></iframe></div> |
+
+The illustrations here are better viewed in viewers than with HTML support (e.g., the vscode built-in viewer).
+
+## Environment Requirements
```
Comment (SamitHuang) [inline examples/instantmesh/README.md:68]: -> Requirements

File: examples/instantmesh/README.md [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+1. To kickstart:
+
+```bash
+pip install -r requirements.txt
+```
+
+2. Inference is tested on the machine with the following specs using 1x NPU:
```
Comment (SamitHuang) [inline examples/instantmesh/README.md:76]: How about training requirements

File: examples/instantmesh/README.md [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+Following the original paper, we used Blender to render multiview frames for a 3D object in `.obj` for training. Typically for overfitting, three 3D objects from the objaverse dataset are used. We rendered 5 arbitral views for each object with the corresponding camera parameters extracted.
+
+### Data Usage for Training
+Following the paper, during training, 3 images are processed with the ViT Image Processor and serving as the model input. Together with another 2 images (in total 5), they are randomly zoomed in and cropped into a fixed size as the ground truth. This can be regarded as a data augmentation during training. The camera parameters for each random transform will be used as the input to the model to infer images and alphas for each view's supervision.
+
+## Performance
+Notice that there is no diffusion model in InstantMesh, therefore the reported training `step/s` here is for each batch iteration.
+
+### Training
+| model name   | stage | precision| batch size |cards | image size  | step/s | frames/s | recompute |
```
Comment (SamitHuang) [inline examples/instantmesh/README.md:124]: put step/s and frames/s to the last two columns

File: examples/instantmesh/README.md (hunk 7 above)
Comment (HaFred) [reply examples/instantmesh/README.md]: done

File: examples/instantmesh/README.md (hunk 8 above)
Comment (HaFred) [reply examples/instantmesh/README.md]: Same. This sentence revised

**Review by SamitHuang — APPROVED** [review]
can attach the training results for better reference

**Comment by HaFred** [comment]
> can attach the training results for better reference

**Review by CaitinZhao — APPROVED** [review]

File: mindone/trainers/train_step.py [hunk 10]
```diff
@@ -101,6 +101,10 @@ def __init__(
     def construct(self, *inputs):
         # compute loss
         weights = self.weights
+
+        # for solving grad err PyNative mode in ms2.4.0: RuntimeErr: value is not supported, valBaseRef
+        self.network.set_inputs(*inputs)
```
Comment (CaitinZhao) [inline mindone/trainers/train_step.py:106]: 这个不要在这里加，在你的代码里加

Comment (HaFred) [reply mindone/trainers/train_step.py]: removed

**Review by CaitinZhao — APPROVED** [review]

**Review by vigo999 — APPROVED** [review]
